File: Examples/HoleEtching/HoleEtching.cpp
#include <SF6O2Etching.hpp>
#include <psMakeHole.hpp>
#include <psProcess.hpp>
#include <psSmartPointer.hpp>
#include <psToSurfaceMesh.hpp>
#include <psUtils.hpp>

int main() {
  using NumericType = double;
  constexpr int D = 3;

  // Domain
  NumericType gridDelta = 0.02;
  NumericType xExtent = 1.0;
  NumericType yExtent = 1.0;

  // Geometry
  NumericType holeRadius = 0.2;
  NumericType topRadius = 0.2;
  NumericType maskHeight = 0.2;
  NumericType taperAngle = 0.;

  // Process
  NumericType processTime = 150;
  NumericType totalEtchantFlux = 4.5e16;
  NumericType totalOxygenFlux = 1e18;
  NumericType totalIonFlux = 2e16;
  NumericType ionEnergy = 100;
  NumericType A_O = 3.;

  // Parse the parameters
  int P, y;

  auto geometry = psSmartPointer<psDomain<NumericType, D>>::New();
  psMakeHole<NumericType, D>(
      geometry, gridDelta /* grid delta */, xExtent /*x extent*/,
      yExtent /*y extent*/, holeRadius /*hole radius*/,
      maskHeight /* mask height*/, taperAngle /* tapering angle in degrees */,
      0 /* base height */, false /* periodic boundary */, true /*create mask*/)
      .apply();

  // use pre-defined model SF6O2 etching model
  auto model = psSmartPointer<SF6O2Etching<NumericType, D>>::New(
      totalIonFlux /*ion flux*/, totalEtchantFlux /*etchant flux*/,
      totalOxygenFlux /*oxygen flux*/, ionEnergy /*min ion energy (eV)*/,
      A_O /*oxy sputter yield*/, 0 /*mask material ID*/);

  // process setup
  psProcess<NumericType, D> process;
  process.setDomain(geometry);
  process.setProcessModel(model);
  process.setMaxCoverageInitIterations(10);
  process.setNumberOfRaysPerPoint(1000);
  process.setProcessDuration(processTime);

  // print initial surface
  geometry->printSurface("initial.vtp");

  // run the process
  process.apply();

  // write collected particle meta data (ion energy distribution) to a file
  process.writeParticleDataLogs("ionEnergyDistribution.txt");

  // print final surface
  geometry->printSurface("final.vtp");
}

File: Examples/HoleEtching/HoleEtching.py
import viennaps3d as vps


def main() -> None:
    grid_delta = 0.02
    x_extent = 1.0
    y_extent = 1.0
    hole_radius = 0.2
    mask_height = 0.2
    taper_angle = 0.0
    total_ion_flux = 2e16
    total_etchant_flux = 4.5e16
    total_oxygen_flux = 1e18
    ion_energy = 100
    oxy_sputter_yield = 3.0
    process_time = 150

    geometry = vps.psDomain()
    vps.psMakeHole(geometry, grid_delta, x_extent, y_extent, hole_radius,
                   mask_height, taper_angle, 0.0, False, True).apply()

    model = vps.SF6O2Etching(total_ion_flux, total_etchant_flux,
                             total_oxygen_flux, ion_energy, oxy_sputter_yield,
                             0)

    process = vps.psProcess()
    process.setDomain(geometry)
    process.setProcessModel(model)
    process.setMaxCoverageInitIterations(10)
    process.setNumberOfRaysPerPoint(1000)
    process.setProcessDuration(process_time)

    geometry.printSurface("initial.vtp")

    process.apply()

    geometry.printSurface("final.vtp")


if __name__ == "__main__":
    main()
